### PR TITLE
Smb hashmap/v4

### PR DESCRIPTION
**Pull request by victorjulien**
https://redmine.openinfosecfoundation.org/issues/5672

### CODE REVIEW

**Comment by victorjulien** [comment]
It would be nice to use a LRU mechanism for the timing out of entries (suggested by @catenacyber elsewhere), but that can happen as a follow up task.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 22397

**Comment by catenacyber** [comment]
I will try to take a look at this next week

**Review by catenacyber — COMMENTED** [review]
Thanks for the work.

CI : ✅
Code : looking at it
Commits segmentation : ok
Commit messages : ok
Git ID set : looks fine for me
CLA : you already contributed :-p
Doc update : not needed
Redmine ticket : 🟠 should this be just a sub ticket of 5672 ?
Rustfmt : I guess there are too many changes to run in smb...
Tests : 🟠 do you have SV tests ?
Dependencies added: none

File: rust/src/smb/smb.rs [hunk 1]
```diff
@@ -1101,6 +1109,21 @@ impl SMBState {
      * was received. */
     fn post_gap_housekeeping(&mut self, dir: Direction)
     {
+        let timeoutts = self.ts - 30;
+        if self.ts_ssn_gap || self.tc_ssn_gap {
+            let _pre_size = self.guid2name_map.len();
+            self.guid2name_map.retain(|_,(_p,ts)| {
```
Comment (catenacyber) [inline rust/src/smb/smb.rs:1114]: Is this iterating all the entries in the (unbounded) hashmap ?..

File: rust/src/smb/smb.rs [hunk 2]
```diff
@@ -1885,22 +1884,26 @@ impl SMBState {
     pub fn parse_tcp_data_ts_gap(&mut self, gap_size: u32) -> AppLayerResult {
         SCLogDebug!("GAP of size {} in toserver direction", gap_size);
         let consumed = self.handle_skip(Direction::ToServer, gap_size);
+        if consumed == gap_size {
```
Comment (catenacyber) [inline rust/src/smb/smb.rs:1887]: Should be squashed into the first commit changing this ;-)

File: rust/src/smb/smb.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/smb/smb.rs:1114]: I meant this looks bad :
If I put 65k elements in this unbounded hash maps, and then have 65k mini gaps (in less than 30 seconds), You get 4G iterations

I think there should be bounds to the number of elements that can be added to this hashmap...

File: rust/src/smb/smb.rs (hunk 2 above)
Comment (victorjulien) [reply rust/src/smb/smb.rs:1887]: Done

File: rust/src/smb/smb.rs (hunk 1 above)
Comment (victorjulien) [reply rust/src/smb/smb.rs:1114]: Using the Lru experiment here #11786 we don't need to do anything anymore. We just set a max size.

**Comment by victorjulien** [comment]
> It would be nice to use a LRU mechanism for the timing out of entries (suggested by @catenacyber elsewhere), but that can happen as a follow up task.

Done in #11786.

File: rust/src/smb/smb.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/smb/smb.rs:1114]: Cool :-)

**Pull request closed without merging**